bson/src/main/org/bson/internal/ProvidersCodecRegistry.java

```diff
@@ -67,19 +68,18 @@ public <T> Codec<T> get(final Class<T> clazz, final CodecRegistry registry) {
     @SuppressWarnings({"unchecked"})
     public <T> Codec<T> get(final ChildCodecRegistry<T> context) {
         CodecCacheKey codecCacheKey = new CodecCacheKey(context.getCodecClass(), context.getTypes().orElse(null));
-        if (!codecCache.containsKey(codecCacheKey)) {
+        return codecCache.<T>get(codecCacheKey).orElseGet(() -> {
             for (CodecProvider provider : codecProviders) {
                 Codec<T> codec = provider.get(context.getCodecClass(), context);
                 if (codec != null) {
                     if (codec instanceof Parameterizable && context.getTypes().isPresent()) {
                         codec = (Codec<T>) ((Parameterizable) codec).parameterize(context, context.getTypes().get());
                     }
-                    return codecCache.putIfMissing(codecCacheKey, codec);
+                    return codecCache.putIfAbsent(codecCacheKey, codec);
```

Comment:
I have just realized that this whole thing can and should be done as a single `compute`. Will do.

Reply:
Turns out, `IllegalStateException: Recursive update` is a thing despite the map being concurrent :( Well, we can't have all the good things.